File: cli/xbin/otel.py
"""xbin OpenTelemetry: auto-instrumentation hooks and environment setup.

Sets up standard OTel environment variables so apps can export traces,
metrics, and logs without configuration.

Environment variables set by the launcher:
  OTEL_SERVICE_NAME        — name of the service (from xbin metadata)
  OTEL_RESOURCE_ATTRIBUTES — service.name, service.version, deployment.mode
  OTEL_EXPORTER_OTLP_ENDPOINT — OTLP endpoint (if --otel-endpoint is set)
  OTEL_EXPORTER_OTLP_PROTOCOL — "grpc" or "http/protobuf"
  OTEL_METRICS_EXPORTER   — "otlp", "prometheus", or "none"
  OTEL_LOGS_EXPORTER      — "otlp", "console", or "none"
  OTEL_TRACES_EXPORTER    — "otlp", "console", or "none"
  OTEL_PYTHON_AUTO_INSTRUMENTATION_ENABLED — "true" (for Python auto-instrument)
"""

from __future__ import annotations

import os
from typing import Any


def build_otel_env(
    service_name: str,
    version: str = "",
    endpoint: str = "",
    protocol: str = "grpc",
    traces_exporter: str = "otlp",
    metrics_exporter: str = "otlp",
    logs_exporter: str = "none",
) -> dict[str, str]:
    """Build OTel environment variables for the launcher to inject.

    Args:
        service_name: name of the service (e.g. "myapp")
        version: service version string
        endpoint: OTLP collector endpoint (e.g. "http://localhost:4317")
        protocol: "grpc" or "http/protobuf"
        traces_exporter: "otlp", "console", or "none"
        metrics_exporter: "otlp", "prometheus", or "none"
        logs_exporter: "otlp", "console", or "none"

    Returns:
        Dict of environment variables to inject.
    """
    env: dict[str, str] = {
        "OTEL_SERVICE_NAME": service_name,
        "OTEL_TRACES_EXPORTER": traces_exporter,
        "OTEL_METRICS_EXPORTER": metrics_exporter,
        "OTEL_LOGS_EXPORTER": logs_exporter,
    }

    resource_attrs = f"service.name={service_name}"
    if version:
        resource_attrs += f",service.version={version}"
    resource_attrs += ",deployment.mode=server"
    env["OTEL_RESOURCE_ATTRIBUTES"] = resource_attrs

    if endpoint:
        env["OTEL_EXPORTER_OTLP_ENDPOINT"] = endpoint
        env["OTEL_EXPORTER_OTLP_PROTOCOL"] = protocol

    # Python auto-instrumentation hook
    if traces_exporter != "none" or metrics_exporter != "none":
        env["OTEL_PYTHON_AUTO_INSTRUMENTATION_ENABLED"] = "true"

    return env


def get_otel_config() -> dict[str, Any]:
    """Read current OTel configuration from environment.

    Returns a dict with all OTel-related env vars, useful for introspection.
    """
    otel_vars = [
        "OTEL_SERVICE_NAME",
        "OTEL_RESOURCE_ATTRIBUTES",
        "OTEL_EXPORTER_OTLP_ENDPOINT",
        "OTEL_EXPORTER_OTLP_PROTOCOL",
        "OTEL_TRACES_EXPORTER",
        "OTEL_METRICS_EXPORTER",
        "OTEL_LOGS_EXPORTER",
        "OTEL_PYTHON_AUTO_INSTRUMENTATION_ENABLED",
    ]
    config: dict[str, Any] = {}
    for var in otel_vars:
        val = os.environ.get(var)
        if val is not None:
            config[var] = val
    return config


def format_resource_attributes(attrs_str: str) -> dict[str, str]:
    """Parse OTEL_RESOURCE_ATTRIBUTES string into a dict.

    Format: "key1=value1,key2=value2"
    """
    result: dict[str, str] = {}
    if not attrs_str:
        return result
    for pair in attrs_str.split(","):
        if "=" in pair:
            key, _, value = pair.partition("=")
            result[key.strip()] = value.strip()
    return result

File: xbin-core/src/otel.rs
use std::collections::HashMap;
use std::env;

pub fn build_otel_env(
    service_name: &str,
    version: &str,
    endpoint: &str,
    protocol: &str,
    traces_exporter: &str,
    metrics_exporter: &str,
    logs_exporter: &str,
) -> HashMap<String, String> {
    let mut env = HashMap::new();

    env.insert("OTEL_SERVICE_NAME".to_string(), service_name.to_string());
    env.insert(
        "OTEL_TRACES_EXPORTER".to_string(),
        traces_exporter.to_string(),
    );
    env.insert(
        "OTEL_METRICS_EXPORTER".to_string(),
        metrics_exporter.to_string(),
    );
    env.insert("OTEL_LOGS_EXPORTER".to_string(), logs_exporter.to_string());

    let mut resource_attrs = format!("service.name={service_name}");
    if !version.is_empty() {
        resource_attrs.push_str(&format!(",service.version={version}"));
    }
    resource_attrs.push_str(",deployment.mode=server");
    env.insert("OTEL_RESOURCE_ATTRIBUTES".to_string(), resource_attrs);

    if !endpoint.is_empty() {
        env.insert(
            "OTEL_EXPORTER_OTLP_ENDPOINT".to_string(),
            endpoint.to_string(),
        );
        env.insert(
            "OTEL_EXPORTER_OTLP_PROTOCOL".to_string(),
            protocol.to_string(),
        );
    }

    if traces_exporter != "none" || metrics_exporter != "none" {
        env.insert(
            "OTEL_PYTHON_AUTO_INSTRUMENTATION_ENABLED".to_string(),
            "true".to_string(),
        );
    }

    env
}

pub fn get_otel_config() -> HashMap<String, String> {
    let otel_vars = [
        "OTEL_SERVICE_NAME",
        "OTEL_RESOURCE_ATTRIBUTES",
        "OTEL_EXPORTER_OTLP_ENDPOINT",
        "OTEL_EXPORTER_OTLP_PROTOCOL",
        "OTEL_TRACES_EXPORTER",
        "OTEL_METRICS_EXPORTER",
        "OTEL_LOGS_EXPORTER",
        "OTEL_PYTHON_AUTO_INSTRUMENTATION_ENABLED",
    ];
    let mut config = HashMap::new();
    for var in otel_vars {
        if let Ok(val) = env::var(var) {
            config.insert(var.to_string(), val);
        }
    }
    config
}

pub fn format_resource_attributes(attrs_str: &str) -> HashMap<String, String> {
    let mut result = HashMap::new();
    if attrs_str.is_empty() {
        return result;
    }
    for pair in attrs_str.split(',') {
        if let Some((key, value)) = pair.split_once('=') {
            result.insert(key.trim().to_string(), value.trim().to_string());
        }
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_build_otel_env_minimal() {
        let env = build_otel_env("myapp", "", "", "grpc", "otlp", "otlp", "none");
        assert_eq!(env.get("OTEL_SERVICE_NAME").unwrap(), "myapp");
        assert_eq!(env.get("OTEL_TRACES_EXPORTER").unwrap(), "otlp");
        assert_eq!(env.get("OTEL_METRICS_EXPORTER").unwrap(), "otlp");
        assert_eq!(env.get("OTEL_LOGS_EXPORTER").unwrap(), "none");
        let attrs = env.get("OTEL_RESOURCE_ATTRIBUTES").unwrap();
        assert!(attrs.contains("service.name=myapp"));
        assert!(attrs.contains("deployment.mode=server"));
        assert!(!attrs.contains("service.version"));
        assert!(!env.contains_key("OTEL_EXPORTER_OTLP_ENDPOINT"));
        assert_eq!(
            env.get("OTEL_PYTHON_AUTO_INSTRUMENTATION_ENABLED").unwrap(),
            "true"
        );
    }

    #[test]
    fn test_build_otel_env_with_endpoint() {
        let env = build_otel_env(
            "svc",
            "1.0",
            "http://localhost:4317",
            "grpc",
            "otlp",
            "otlp",
            "none",
        );
        assert_eq!(
            env.get("OTEL_EXPORTER_OTLP_ENDPOINT").unwrap(),
            "http://localhost:4317"
        );
        assert_eq!(env.get("OTEL_EXPORTER_OTLP_PROTOCOL").unwrap(), "grpc");
        let attrs = env.get("OTEL_RESOURCE_ATTRIBUTES").unwrap();
        assert!(attrs.contains("service.version=1.0"));
    }

    #[test]
    fn test_format_resource_attributes() {
        let result = format_resource_attributes("a=1,b=2,c=3");
        assert_eq!(result.get("a").unwrap(), "1");
        assert_eq!(result.get("b").unwrap(), "2");
        assert_eq!(result.get("c").unwrap(), "3");
    }

    #[test]
    fn test_format_resource_attributes_empty() {
        assert!(format_resource_attributes("").is_empty());
    }
}
